sql: panic on CREATE INVERTED INDEX

**BUG REPORT**

```
I180227 15:58:05.585150 1 util/log/clog.go:1040  [config] file created at: 2018/02/27 15:58:05
I180227 15:58:05.585150 1 util/log/clog.go:1040  [config] running on machine: Laurens-MBP
I180227 15:58:05.585150 1 util/log/clog.go:1040  [config] binary: CockroachDB CCL v2.0-alpha.20180129-343-gbb1ce80a8-dirty (darwin amd64, built 2018/02/06 16:02:15, go1.9.2)
I180227 15:58:05.585150 1 util/log/clog.go:1040  [config] arguments: [cockroach start --insecure --host=localhost]
```

Created table with JSONB column and tried to create an inverted index.

1. Start three nodes locally, insecure.
2. Create database:
```
CREATE DATABASE json_test;
```
3. Create table:
```
CREATE TABLE users (                                                                                                                  profile_id UUID PRIMARY KEY DEFAULT gen_random_uuid(),                                                                                                   last_updated TIMESTAMP DEFAULT now(),                                                                                                                    user_profile JSONB                                                                                                                                     );
```
4. Insert data:
```
INSERT INTO users (user_profile) VALUES                                                                                               ('{"first_name": "Lola", "last_name": "Dog", "location": "NYC", "online" : true, "friends" : 547}'),                                                     ('{"first_name": "Ernie", "status": "Looking for treats", "location" : "Brooklyn"}'                                                                    );
```
5. Create inverted index:
```
create inverted index dogs on users(user_profile);
```
After, I got the following errors were printed to stderr:

```
driver: bad connection
warning: connection lost!
opening new connection: all session settings will be lost
warning: error retrieving the transaction status: dial tcp [::1]:26257: getsockopt: connection refused
warning: connection lost!
opening new connection: all session settings will be lost
warning: error retrieving the database name: dial tcp [::1]:26257: getsockopt: connection refused
root@:26257 ?>
```

And Node 1 and 3 crashed with the following panics: 

```
panic: initialized checksum = 5b91ac8d

goroutine 1452 [running]:
github.com/cockroachdb/cockroach/pkg/roachpb.(*Value).InitChecksum(0xc42673d370, 0xc427f433b0, 0x2c, 0x30)
	/Users/laurenhirata/go/src/github.com/cockroachdb/cockroach/pkg/roachpb/data.go:251 +0x206
github.com/cockroachdb/cockroach/pkg/roachpb.NewInitPut(0xc427f433b0, 0x2c, 0x30, 0xc425408bc0, 0x5, 0x5, 0x0, 0x0, 0x1, 0x8789608, ...)
	/Users/laurenhirata/go/src/github.com/cockroachdb/cockroach/pkg/roachpb/api.go:890 +0x4d
github.com/cockroachdb/cockroach/pkg/internal/client.(*Batch).InitPut(0xc426c21100, 0x5e048a0, 0xc424fcee00, 0x5f0a740, 0xc42671c218, 0xc42673d401)
	/Users/laurenhirata/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/batch.go:454 +0x1e4
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*indexBackfiller).runChunk.func4(0x6ecca00, 0xc42671c1c0, 0xc4280896c0, 0x10001ff01, 0xc425408d0c)
	/Users/laurenhirata/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/indexbackfiller.go:253 +0x105
github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Txn.func1(0x6ecca00, 0xc42671c1c0, 0xc4280896c0, 0xc425408d0c, 0x0, 0x151751ae2ec8aaf0)
	/Users/laurenhirata/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:516 +0x43
github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Exec(0xc4280896c0, 0x6ecca00, 0xc42671c1c0, 0xc428080101, 0xc420262c40, 0x0, 0x0)
	/Users/laurenhirata/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:753 +0x10c
github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Txn(0xc4206ab530, 0x6ecca00, 0xc42671c1c0, 0xc424fced40, 0x0, 0x0)
	/Users/laurenhirata/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:515 +0x14b
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*indexBackfiller).runChunk(0xc426a3d000, 0x6eccac0, 0xc427deef00, 0xc424fcea60, 0x1, 0x1, 0xc425099968, 0x2, 0x8, 0xc425099990, ...)
	/Users/laurenhirata/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/indexbackfiller.go:236 +0x58d
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*backfiller).mainLoop(0xc426a3d000, 0x6eccac0, 0xc427deef00, 0x5f36281, 0x8)
	/Users/laurenhirata/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/backfiller.go:137 +0x5ec
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*backfiller).Run(0xc426a3d000, 0xc424e1a7f8)
	/Users/laurenhirata/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/backfiller.go:89 +0x21a
created by github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Start
	/Users/laurenhirata/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:466 +0x3c5
```

```
heartbeat failed: rpc error: code = Unavailable desc = all SubConns are in TransientFailure [exhausted])
I180227 23:13:30.965400 443 storage/node_liveness.go:710  [n1,hb] retrying liveness update after storage.errRetryLiveness: result is ambiguous (error=initial connection heartbeat failed: rpc error: code = Unavailable desc = all SubConns are in TransientFailure [exhausted])
I180227 23:13:30.966840 443 storage/node_liveness.go:710  [n1,hb] retrying liveness update after storage.errRetryLiveness: result is ambiguous (error=initial connection heartbeat failed: rpc error: code = Unavailable desc = all SubConns are in TransientFailure [exhausted])
I180227 23:13:30.969412 443 storage/node_liveness.go:710  [n1,hb] retrying liveness update after storage.errRetryLiveness: result is ambiguous (error=initial connection heartbeat failed: rpc error: code = Unavailable desc = all SubConns are in TransientFailure [exhausted])
I180227 23:13:30.972162 443 storage/node_liveness.go:710  [n1,hb] retrying liveness update after storage.errRetryLiveness: result is ambiguous (erpanic: initialized checksum = 5b91ac8d

goroutine 102031 [running]:
github.com/cockroachdb/cockroach/pkg/roachpb.(*Value).InitChecksum(0xc430f87370, 0xc42f4560c0, 0x2c, 0x30)
	/Users/laurenhirata/go/src/github.com/cockroachdb/cockroach/pkg/roachpb/data.go:251 +0x206
github.com/cockroachdb/cockroach/pkg/roachpb.NewInitPut(0xc42f4560c0, 0x2c, 0x30, 0xc42c4f3fa0, 0x5, 0x5, 0x0, 0x0, 0x1, 0xfd319a8, ...)
	/Users/laurenhirata/go/src/github.com/cockroachdb/cockroach/pkg/roachpb/api.go:890 +0x4d
github.com/cockroachdb/cockroach/pkg/internal/client.(*Batch).InitPut(0xc42a4ef100, 0x5e048a0, 0xc431d485c0, 0x5f0a740, 0xc42eb070d8, 0xc430f87401)
	/Users/laurenhirata/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/batch.go:454 +0x1e4
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*indexBackfiller).runChunk.func4(0x6ecca00, 0xc42eb07080, 0xc428ff2d00, 0x10005cc01, 0xc42c43e0cc)
	/Users/laurenhirata/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/indexbackfiller.go:253 +0x105
github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Txn.func1(0x6ecca00, 0xc42eb07080, 0xc428ff2d00, 0xc42c43e0cc, 0x0, 0x151751b0e41758c8)
	/Users/laurenhirata/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:516 +0x43
github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Exec(0xc428ff2d00, 0x6ecca00, 0xc42eb07080, 0xc428ff0101, 0xc42c24cfe0, 0x0, 0x0)
	/Users/laurenhirata/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:753 +0x10c
github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Txn(0xc420878870, 0x6ecca00, 0xc42eb07080, 0xc431d483c0, 0x0, 0x0)
	/Users/laurenhirata/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:515 +0x14b
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*indexBackfiller).runChunk(0xc43afe8800, 0x6eccac0, 0xc431bb3ec0, 0xc431cd3be0, 0x1, 0x1, 0xc42c4f3c88, 0x2, 0x8, 0xc42c4f3c90, ...)
	/Users/laurenhirata/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/indexbackfiller.go:236 +0x58d
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*backfiller).mainLoop(0xc43afe8800, 0x6eccac0, 0xc431bb3ec0, 0x5f36281, 0x8)
	/Users/laurenhirata/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/backfiller.go:137 +0x5ec
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*backfiller).Run(0xc43afe8800, 0xc42bf887f8)
	/Users/laurenhirata/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/backfiller.go:89 +0x21a
created by github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Start
	/Users/laurenhirata/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:466 +0x3c5
```

--- a/pkg/sql/distsqlrun/indexbackfiller.go
+++ b/pkg/sql/distsqlrun/indexbackfiller.go
@@ -180,6 +180,12 @@ func (ib *indexBackfiller) runChunk(
 			if err := sqlbase.EncDatumRowToDatums(ib.types, ib.rowVals, encRow, &ib.da); err != nil {
 				return nil, err
 			}
+
+			// We're resetting the length of this slice for variable length indexes such as inverted
+			// indexes which can append entries to the end of the slice. If we don't do this, then everything
+			// EncodeSecondaryIndexes appends to secondaryIndexEntries for a row, would stay in the slice for
+			// subsequent rows and we would then have duplicates in entries on output.
+			secondaryIndexEntries = secondaryIndexEntries[:len(mutations)]
 			if secondaryIndexEntries, err = sqlbase.EncodeSecondaryIndexes(
 				&ib.spec.Table, added, ib.colIdxMap,
 				ib.rowVals, secondaryIndexEntries); err != nil {

--- a/pkg/sql/logictest/testdata/logic_test/inverted_index
+++ b/pkg/sql/logictest/testdata/logic_test/inverted_index
@@ -458,3 +458,27 @@ SELECT * from d where b @> 'null' ORDER BY a;
 ----
 11  null
 27  [true, false, null, 1.23, "a"]
+
+statement ok
+CREATE TABLE users (
+  profile_id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
+  last_updated TIMESTAMP DEFAULT now(),
+  user_profile JSONB
+);
+
+statement ok
+INSERT INTO users (user_profile) VALUES  ('{"first_name": "Lola", "last_name": "Dog", "location": "NYC", "online" : true, "friends" : 547}'),
+                                         ('{"first_name": "Ernie", "status": "Looking for treats", "location" : "Brooklyn"}');
+
+statement ok
+CREATE INVERTED INDEX dogs on users(user_profile);
+
+query T
+SELECT user_profile from users where user_profile @> '{"first_name":"Lola"}';
+----
+{"first_name": "Lola", "friends": 547, "last_name": "Dog", "location": "NYC", "online": true}
+
+query T
+SELECT user_profile from users where user_profile @> '{"first_name":"Ernie"}';
+----
+ {"first_name": "Ernie", "location": "Brooklyn", "status": "Looking for treats"}

--- a/pkg/sql/sqlbase/table.go
+++ b/pkg/sql/sqlbase/table.go
@@ -1709,16 +1709,20 @@ func EncodeSecondaryIndexes(
 	values []tree.Datum,
 	secondaryIndexEntries []IndexEntry,
 ) ([]IndexEntry, error) {
+	if len(secondaryIndexEntries) != len(indexes) {
+		panic("Length of secondaryIndexEntries is not equal to the number of indexes.")
+	}
 	for i := range indexes {
 		entries, err := EncodeSecondaryIndex(tableDesc, &indexes[i], colMap, values)
 		if err != nil {
 			return secondaryIndexEntries, err
 		}
+		secondaryIndexEntries[i] = entries[0]
+
+		// This is specifically for inverted indexes which can have more than one entry
+		// associated with them.
 		if len(entries) > 1 {
 			secondaryIndexEntries = append(secondaryIndexEntries, entries[1:]...)
-		}
-		if len(entries) > 0 {
-			secondaryIndexEntries[i] = entries[0]
 		}
 	}
 	return secondaryIndexEntries, nil

